File: scripts/docs/find-prd-triggers.py
#!/usr/bin/env python3
from __future__ import annotations

import json
import subprocess
import sys
from pathlib import Path


ROOT = Path(__file__).resolve().parents[2]


def git(*args: str) -> str:
    result = subprocess.run(
        ["git", *args],
        cwd=ROOT,
        capture_output=True,
        text=True,
        check=True,
    )
    return result.stdout


def parse_frontmatter(blob: str) -> dict[str, str]:
    lines = blob.splitlines()
    if not lines or lines[0].strip() != "---":
        return {}
    data: dict[str, str] = {}
    idx = 1
    while idx < len(lines):
        line = lines[idx]
        if line.strip() == "---":
            return data
        if ":" in line:
            key, value = line.split(":", 1)
            data[key.strip()] = value.strip().strip("'\"")
        idx += 1
    return {}


def show(ref: str, path: str) -> str:
    try:
        return git("show", f"{ref}:{path}")
    except subprocess.CalledProcessError:
        return ""


def list_changes(base: str, head: str) -> list[tuple[str, str | None, str]]:
    output = git("diff", "--name-status", "--find-renames", base, head, "--", "docs/planning")
    changes: list[tuple[str, str | None, str]] = []
    for line in output.splitlines():
        if not line.strip():
            continue
        parts = line.split("\t")
        status = parts[0]
        if status.startswith("R"):
            changes.append((status, parts[1], parts[2]))
        elif status.startswith(("A", "M", "C")):
            changes.append((status, None, parts[1]))
    return changes


def trigger_reason(status_code: str, old_path: str | None, new_path: str, base: str, head: str) -> str | None:
    new_meta = parse_frontmatter(show(head, new_path))
    if not new_meta:
        return None
    if "/approved/" not in new_path:
        return None
    if new_meta.get("status") != "approved":
        return None
    if new_meta.get("prd_mode") not in {"generate", "update"}:
        return None

    if status_code.startswith("R") and old_path and "/proposed/" in old_path and "/approved/" in new_path:
        return "promotion"
    if status_code.startswith("A"):
        return "new-approved-doc"

    old_blob = show(base, old_path or new_path)
    old_meta = parse_frontmatter(old_blob)
    old_status = old_meta.get("status")

    if old_status != "approved":
        return "status-change"
    if new_meta.get("prd_mode") == "update":
        return "approved-doc-updated"
    return None


def main() -> int:
    if len(sys.argv) != 3:
        print("usage: find-prd-triggers.py <base-ref> <head-ref>", file=sys.stderr)
        return 2

    base, head = sys.argv[1], sys.argv[2]
    events = []
    for status_code, old_path, new_path in list_changes(base, head):
        reason = trigger_reason(status_code, old_path, new_path, base, head)
        if not reason:
            continue
        meta = parse_frontmatter(show(head, new_path))
        events.append(
            {
                "reason": reason,
                "doc_path": new_path,
                "doc_id": meta.get("id", ""),
                "title": meta.get("title", ""),
                "prd_mode": meta.get("prd_mode", ""),
                "target_branch": meta.get("target_branch", ""),
                "linked_prd": meta.get("linked_prd", ""),
                "source_path": old_path or new_path,
            }
        )

    print(json.dumps(events, indent=2))
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: scripts/docs/find-prd-triggers.mjs
#!/usr/bin/env node
import { execFileSync } from "node:child_process";
import path from "node:path";
import { fileURLToPath } from "node:url";

const root = path.resolve(path.dirname(fileURLToPath(import.meta.url)), "..", "..");

function git(...args) {
  return execFileSync("git", args, {
    cwd: root,
    encoding: "utf8",
  });
}

function parseFrontmatter(blob) {
  const lines = blob.split(/\r?\n/);
  if (lines.length === 0 || lines[0].trim() !== "---") {
    return {};
  }
  const data = {};
  for (let index = 1; index < lines.length; index += 1) {
    const line = lines[index];
    if (line.trim() === "---") {
      return data;
    }
    if (!line.includes(":")) {
      continue;
    }
    const [rawKey, ...rest] = line.split(":");
    data[rawKey.trim()] = rest.join(":").trim().replace(/^['"]|['"]$/g, "");
  }
  return {};
}

function show(ref, filePath) {
  try {
    return git("show", `${ref}:${filePath}`);
  } catch {
    return "";
  }
}

function listChanges(base, head) {
  const output = git("diff", "--name-status", "--find-renames", base, head, "--", "docs/planning");
  const changes = [];
  for (const line of output.split(/\r?\n/)) {
    if (!line.trim()) {
      continue;
    }
    const parts = line.split("\t");
    const status = parts[0];
    if (status.startsWith("R")) {
      changes.push([status, parts[1], parts[2]]);
    } else if (/^[AMC]/.test(status)) {
      changes.push([status, null, parts[1]]);
    }
  }
  return changes;
}

function triggerReason(statusCode, oldPath, newPath, base, head) {
  const newMeta = parseFrontmatter(show(head, newPath));
  if (Object.keys(newMeta).length === 0) {
    return null;
  }
  if (!newPath.includes("/approved/")) {
    return null;
  }
  if (newMeta.status !== "approved") {
    return null;
  }
  if (!["generate", "update"].includes(newMeta.prd_mode)) {
    return null;
  }

  if (statusCode.startsWith("R") && oldPath && oldPath.includes("/proposed/") && newPath.includes("/approved/")) {
    return "promotion";
  }
  if (statusCode.startsWith("A")) {
    return "new-approved-doc";
  }

  const oldMeta = parseFrontmatter(show(base, oldPath ?? newPath));
  if (oldMeta.status !== "approved") {
    return "status-change";
  }
  if (newMeta.prd_mode === "update") {
    return "approved-doc-updated";
  }
  return null;
}

if (process.argv.length !== 4) {
  console.error("usage: find-prd-triggers.mjs <base-ref> <head-ref>");
  process.exit(2);
}

const [, , base, head] = process.argv;
const events = [];

for (const [statusCode, oldPath, newPath] of listChanges(base, head)) {
  const reason = triggerReason(statusCode, oldPath, newPath, base, head);
  if (!reason) {
    continue;
  }
  const meta = parseFrontmatter(show(head, newPath));
  events.push({
    reason,
    doc_path: newPath,
    doc_id: meta.id ?? "",
    title: meta.title ?? "",
    prd_mode: meta.prd_mode ?? "",
    target_branch: meta.target_branch ?? "",
    linked_prd: meta.linked_prd ?? "",
    source_path: oldPath ?? newPath,
  });
}

console.log(JSON.stringify(events, null, 2));
